com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/ScopesRequestHandler.java

```diff
@@ -39,14 +39,14 @@
     public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response, IDebugAdapterContext context) {
         ScopesArguments scopesArgs = (ScopesArguments) arguments;
         List<Types.Scope> scopes = new ArrayList<>();
-        JdiObjectProxy<StackFrame> stackFrameProxy = (JdiObjectProxy<StackFrame>) context.getRecyclableIdPool().getObjectById(scopesArgs.frameId);
-        if (stackFrameProxy == null) {
+        StackFrameReference stackFrameReference = (StackFrameReference) context.getRecyclableIdPool().getObjectById(scopesArgs.frameId);
+        if (stackFrameReference == null) {
             response.body = new Responses.ScopesResponseBody(scopes);
             return CompletableFuture.completedFuture(response);
         }
-        StackFrame stackFrame = stackFrameProxy.getProxiedObject();
-        VariableProxy localScope = new VariableProxy(stackFrame.thread().uniqueID(), "Local", stackFrame);
-        int localScopeId = context.getRecyclableIdPool().addObject(stackFrame.thread().uniqueID(), localScope);
+        ThreadReference thread = stackFrameReference.getThread();
```

Comment:
Here's the problem. If someone has the thread instance, then he is able to access the stack frames directly. The stack frame manager is bypassed this way.

Reply:
I'm thinking whether putting a lock on the stack frame is enough. Maybe the lock should be on a thread.